otherlibs/unix/link.c

```diff
@@ -18,18 +18,32 @@
 #include <caml/signals.h>
 #include "unixsupport.h"
 
-CAMLprim value unix_link(value path1, value path2)
+#include <fcntl.h>
+#include <unistd.h>
+
+CAMLprim value unix_link(value follow, value path1, value path2)
 {
-  CAMLparam2(path1, path2);
+  CAMLparam3(follow, path1, path2);
   char * p1;
   char * p2;
   int ret;
   caml_unix_check_path(path1, "link");
   caml_unix_check_path(path2, "link");
   p1 = caml_strdup(String_val(path1));
   p2 = caml_strdup(String_val(path2));
+  ret =
+    Is_block(follow) && Bool_val(Field(follow, 0)) /* Some true */
+    ? AT_SYMLINK_FOLLOW
+    : 0;
   caml_enter_blocking_section();
-  ret = link(p1, p2);
+  if (follow == Val_int(0) /* None */)
+    ret = link(p1, p2);
+  else /* Some bool */
+# if _XOPEN_VERSION >= 700 || _POSIX_VERSION >= 200809L
+    ret = linkat(AT_FDCWD, p1, AT_FDCWD, p2, ret);
```

Comment:
That's a very good point (I don't think the access in the if is correct, therefore, either) - it would be better to have two lots of enter_blocking_section around both link and linkat (which again, begs my previous question of why this isn't just a separate binding).